### PR TITLE
Move SystemLog plugin into neo-cli as a native logger with on/off functionalities.

### PR DESCRIPTION
**Pull request by joeqian10**
This fixes neo-modules issue [228](https://github.com/neo-project/neo-modules/issues/228).

### CODE REVIEW

File: neo-cli/CLI/MainService.Node.cs [hunk 1]
```diff
@@ -86,5 +86,23 @@ private void OnShowStateCommand()
             Console.WriteLine();
             Console.CursorVisible = true;
         }
+
+        /// <summary>
+        /// Start the default ILogPlugin
+        /// </summary>
+        [ConsoleCommand("log on", Category = "Node Commands")]
+        private void OnLogOnCommand()
```
Comment (erikzhang) [inline neo-cli/CLI/MainService.Node.cs:94]: Why do we need these commands?

File: neo-cli/CLI/MainService.cs [hunk 2]
```diff
@@ -65,6 +66,8 @@ private set
         protected override string Prompt => "neo";
         public override string ServiceName => "NEO-CLI";
 
+        private readonly Logger logger = new Logger();
```
Comment (erikzhang) [inline neo-cli/CLI/MainService.cs:69]: The plugin should be initialized in the `Start()` method. (After the initialization of `NeoSystem`)

File: neo-cli/SystemLog/Logger.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.IO;
+using static System.IO.Path;
+using Microsoft.Extensions.Configuration;
+
+namespace Neo.SystemLog
+{
+    public class Logger : Plugin, ILogPlugin
```
Comment (erikzhang) [inline neo-cli/SystemLog/Logger.cs:11]: You also need to override `ConfigFile` and `Path`.

File: neo-cli/SystemLog/ConsoleColorSet.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Neo.SystemLog
+{
+    internal class ConsoleColorSet
```
Comment (shargon) [inline neo-cli/SystemLog/ConsoleColorSet.cs:7]: Move this to https://github.com/neo-project/neo-node/tree/master/Neo.ConsoleService ?

File: neo-cli/CLI/MainService.Node.cs (hunk 1 above)
Comment (joeqian10) [reply neo-cli/CLI/MainService.Node.cs:94]: We need these to toggle log function when neo-cli is running as @shargon mentioned. Do you have any better ideas?

File: neo-cli/CLI/MainService.cs (hunk 2 above)
Comment (joeqian10) [reply neo-cli/CLI/MainService.cs]: I have moved the initialization into `Start()` method.

File: neo-cli/CLI/MainService.cs [hunk 5]
```diff
@@ -65,6 +66,8 @@ private set
         protected override string Prompt => "neo";
         public override string ServiceName => "NEO-CLI";
 
+        private Logger logger;
```
Comment (shargon) [inline neo-cli/CLI/MainService.cs:69]:
```suggestion
        private readonly Logger logger;
```

File: neo-cli/SystemLog/Logger.cs (hunk 3 above)
Comment (joeqian10) [reply neo-cli/SystemLog/Logger.cs:12]: I have overriden `ConfigFile` but `Path` is not marked as `virtual`.

File: neo-cli/CLI/MainService.cs (hunk 5 above)
Comment (erikzhang) [reply neo-cli/CLI/MainService.cs:69]: It can't be `readonly` because it is not initialized in constructor.

Comment (joeqian10) [reply neo-cli/CLI/MainService.cs:69]: If you mark it as `readonly`, a readonly field cannot be assigned to (except in a constructor of the class in which the field is defined or a variable initializer). And I have put the initialization in the `Start()` method.

File: neo-cli/CLI/MainService.Node.cs (hunk 1 above)
Comment (erikzhang) [reply neo-cli/CLI/MainService.Node.cs:94]: I mean why do we need to toggle log function when neo-cli is running? Maybe a command-line argument is enough. Example: `--log-on` or `--log-off`.

Comment (shargon) [reply neo-cli/CLI/MainService.Node.cs:94]: If we have the config inside the json, I think that it's not needed

File: neo-cli/SystemLog/Logger.cs (hunk 3 above)
Comment (erikzhang) [reply neo-cli/SystemLog/Logger.cs:12]: It should be. We need to modify the `Path` to be `virtual`.

File: neo-cli/SystemLog/ConsoleColorSet.cs (hunk 4 above)
Comment (joeqian10) [reply neo-cli/SystemLog/ConsoleColorSet.cs:7]: Sure, no problem.

File: neo-cli/CLI/MainService.Node.cs (hunk 1 above)
Comment (joeqian10) [reply neo-cli/CLI/MainService.Node.cs:94]: OK, I will add a configuration section inside `config.json`.

File: neo-cli/CLI/Logger.cs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+                if (!string.IsNullOrEmpty(Settings.Default.Logger.Path))
+                {
+                    StringBuilder sb = new StringBuilder(source);
+                    foreach (char c in GetInvalidFileNameChars())
+                        sb.Replace(c, '-');
+                    var path = Combine(Settings.Default.Logger.Path, sb.ToString());
+                    Directory.CreateDirectory(path);
+                    path = Combine(path, $"{now:yyyy-MM-dd}.log");
+                    File.AppendAllLines(path, new[] { $"[{level}]{log}" });
```
Comment (superboyiii) [inline neo-cli/CLI/Logger.cs:87]:
```suggestion
                    try
                    {
                        File.AppendAllLines(path, new[] { $"[{level}]{log}" });
                    }
                    catch (IOException)
                    {
                        Console.WriteLine("Error writing the log file: " + path);

                    }
```

**Comment by superboyiii** [comment]
@joeqian10 Unhandled exceptions are not recorded in System log.
image
Now:
image
Before:
image

File: neo-cli/CLI/Logger.cs [hunk 7]
```diff
@@ -0,0 +1,100 @@
+using Neo.Plugins;
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.IO;
+using static System.IO.Path;
+using Microsoft.Extensions.Configuration;
+using System.Reflection;
+using Neo.ConsoleService;
```
Comment (superboyiii) [inline neo-cli/CLI/Logger.cs:9]:
```suggestion
using Neo.Plugins;
using System;
using System.Text;
using System.IO;
using static System.IO.Path;
using System.Reflection;
using Neo.ConsoleService;
```
Delete unused dependency.

**Comment by erikzhang** [comment]
Why merge it so quickly? I didn't finish my review.

**Comment by erikzhang** [comment]
Reverted.

**Pull request merged**